src/main/cpp/js.cpp

```diff
@@ -0,0 +1,1071 @@
+            }
+            else if (c == u8'\\') {
+                escaped = true;
+            }
+            else if (c == u8'`') {
+                return String_Literal_Result { .length = length + 1,
+                                               .is_template_literal = true,
+                                               .terminated = true,
+                                               .has_substitution = has_substitution };
+            }
+            else if (c == u8'$' && length + 1 < str.length() && str[length + 1] == u8'{') {
+                has_substitution = true;
+                const std::size_t subst_length = match_template_substitution(str.substr(length));
+                if (subst_length == 0) { // Unterminated.
+                    return String_Literal_Result { .length = str.length(),
+                                                   .is_template_literal = true,
+                                                   .terminated = false,
+                                                   .has_substitution = true };
+                }
+                length += subst_length;
+                continue;
+            }
+
+            ++length;
+        }
+
+        // Unterminated template literal.
+        return String_Literal_Result { .length = length,
+                                       .is_template_literal = true,
+                                       .terminated = false,
+                                       .has_substitution = has_substitution };
+    }
+
+    if (str[0] == u8'"' || str[0] == u8'\'') {
+        const char8_t quote = str[0];
+        std::size_t length = 1;
+        bool escaped = false;
+
+        while (length < str.length()) {
+            const char8_t c = str[length];
+
+            if (escaped) {
+                escaped = false;
+            }
+            else if (c == u8'\\') {
+                escaped = true;
+            }
+            else if (c == quote) {
+                return String_Literal_Result { .length = length + 1,
+                                               .is_template_literal = false,
+                                               .terminated = true,
+                                               .has_substitution = false };
+            }
+            else if (c == u8'\n') {
+                return String_Literal_Result { .length = length,
+                                               .is_template_literal = false,
+                                               .terminated = false,
+                                               .has_substitution = false };
+            }
+
+            ++length;
+        }
+
+        return String_Literal_Result { .length = length,
+                                       .is_template_literal = false,
+                                       .terminated = false,
+                                       .has_substitution = false };
+    }
+
+    return {};
+}
+
+std::size_t match_template_substitution(std::u8string_view str)
+{
```

Comment:
```suggestion
{
    // https://262.ecma-international.org/15.0/index.html#sec-template-literal-lexical-components
```
On another note, it's not actually clear to me if you need this function at all.

If the highlighter sees a high quote, it can just commit to matching template literals, regardless whether this `match` function succeeds. The highlighter needs to go over everything as second time anyway so the contents of templates are properly highlighted, so I think this could just be eliminated.